### ISSUE TITLE
Realm: GCC 9 build error

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1296]
In the attempt to upgrade CI to 20.04 (#1295), I noticed this build error on GCC 9 on 32-bit architectures:

```
/builds/StanfordLegion/legion/runtime/realm/idx_impl.cc: In instantiation of ‘Realm::IndexSpaceGeneric::IndexSpaceGeneric(const Realm::IndexSpace<N, T>&) [with int N = 2; T = long long int]’:
/builds/StanfordLegion/legion/runtime/realm/idx_impl.cc:197:3:   required from here
/builds/StanfordLegion/legion/runtime/realm/idx_impl.cc:83:16: error: placement new constructing an object of type ‘Realm::IndexSpaceGenericImplTyped<2, long long int>’ and size ‘44’ in a region of type ‘Realm::IndexSpaceGeneric::Storage_aligned’ {aka ‘char [32]’} and size ‘32’ [-Werror=placement-new=]
   83 |     impl = new(raw_storage) IndexSpaceGenericImplTyped<N,T>(copy_from);
      |                ^~~~~~~~~~~
```

Job log: https://gitlab.com/StanfordLegion/legion/-/jobs/2699425580 ([archived](https://github.com/StanfordLegion/legion/files/9077262/job.log))

**Comment by lightsighter** [comment]
Try changing this line here:

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/indexspace.h#L589

to instead be:

```
static const size_t STORAGE_BYTES = (2*REALM_MAX_DIM + 2) * sizeof(long long);
```

**Comment by elliottslaughter** [comment]
Is this `long long` really a `coord_t`?

Should I be touching any of the surrounding lines? (E.g. https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/indexspace.h#L591)

**Comment by lightsighter** [comment]
> Is this long long really a coord_t?

`coord_t` is only a Legion type now. It doesn't exist in Realm anymore. Really we should be picking the largest type out of this list:

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/indexspace.h#L41

> Should I be touching any of the surrounding lines? (E.g. https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/indexspace.h#L591)

You probably also need to update 591 to get the alignment right too.

**Comment by elliottslaughter** [comment]
Ok, good. As long as there's a principled reason for it to be safe to use `long long`, I'm fine with it.

**Comment by elliottslaughter** [comment]
This seems to fail with:

```
/repo/runtime/realm/dynamic_templates.inl:215:55: error: ‘demux<Realm::DynamicTemplates::ListProduct2<L1, L2>::DemuxHelper2<TARGET, T1>, A1>’ is not a member of ‘Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >’
/repo/runtime/realm/dynamic_templates.inl: In instantiation of ‘static void Realm::DynamicTemplates::ListProduct2<L1, L2>::DemuxHelper1<TARGET>::demux(Realm::DynamicTemplates::TagType, A1) [with T1 = Realm::DynamicTemplates::Int<3>; A1 = int*; TARGET = Legion::Internal::NT_TemplateHelper::DimHelper; L1 = Realm::DynamicTemplates::IntList<1, 4>; L2 = Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >; Realm::DynamicTemplates::TagType = unsigned int]’:
/repo/runtime/realm/dynamic_templates.h:209:53:   required from ‘static void Realm::DynamicTemplates::IntDemuxHelper<TARGET, BASE, DELTA>::demux(int, T1, T2) [with T1 = unsigned int; T2 = int*; TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > > >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; int BASE = 1; int DELTA = 2]’
/repo/runtime/realm/dynamic_templates.h:211:71:   required from ‘static void Realm::DynamicTemplates::IntDemuxHelper<TARGET, BASE, DELTA>::demux(int, T1, T2) [with T1 = unsigned int; T2 = int*; TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > > >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; int BASE = 1; int DELTA = 3]’
/repo/runtime/realm/dynamic_templates.inl:154:64:   required from ‘static void Realm::DynamicTemplates::IntList<MIN, MAX>::demux(int, T1, T2) [with TARGET = Realm::DynamicTemplates::ListProduct2<Realm::DynamicTemplates::IntList<1, 4>, Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >  >::DemuxHelper1<Legion::Internal::NT_TemplateHelper::DimHelper>; T1 = unsigned int; T2 = int*; int MIN = 1; int MAX = 4]’
/repo/runtime/realm/dynamic_templates.inl:241:60:   required from ‘static void Realm::DynamicTemplates::ListProduct2<L1, L2>::demux(Realm::DynamicTemplates::TagType, A1) [with TARGET = Legion::Internal::NT_TemplateHelper::DimHelper; A1 = int*; L1 = Realm::DynamicTemplates::IntList<1, 4>; L2 = Realm::DynamicTemplates::TypeListElem<int, Realm::DynamicTemplates::TypeListElem<unsigned int, Realm::DynamicTemplates::TypeListElem<long long int, Realm::DynamicTemplates::TypeListTerm> > >; Realm::DynamicTemplates::TagType = unsigned int]’
/repo/runtime/legion/legion_types.h:1893:43:   required from here
```

**Comment by lightsighter** [comment]
Ok, let's try with this one because gcc is stupider than clang. This one has at least been tested with GCC 9 and clang 10:

```
diff --git a/runtime/realm/dynamic_templates.h b/runtime/realm/dynamic_templates.h
index 1622976af..8e97ac835 100644
--- a/runtime/realm/dynamic_templates.h
+++ b/runtime/realm/dynamic_templates.h
@@ -85,6 +85,20 @@ namespace Realm {
        typedef _HEAD TYPE;
       };
 
+      // MaxSize support
+      struct MaxSize {
+        // Unforuntately std::max is not constexpr until c++14
+        static constexpr size_t value = sizeof(_HEAD) < _TAIL::MaxSize::value ?
+          _TAIL::MaxSize::value : sizeof(_HEAD);
+      };
+
+      // MaxSizeType support
+      template <size_t SIZE>
+      struct MaxSizeType {
+        typedef typename std::conditional<SIZE == sizeof(_HEAD),_HEAD,
+                  typename _TAIL::template MaxSizeType<SIZE> >::type TYPE;
+      };
+
       // demux support
       template <typename TARGET, int N>
       struct DemuxHelper {
@@ -138,6 +152,18 @@ namespace Realm {
        static const int INDEX = ERROR_INDEX_NOT_IN_LIST<ORIG>::TYPE;
       };
 
+      // MaxSize support
+      struct MaxSize {
+        static constexpr size_t value = 0;
+      };
+
+      struct ERROR_SIZE_NOT_IN_LIST { };
+
+      template <size_t SIZE>
+      struct MaxSizeType {
+        typedef ERROR_SIZE_NOT_IN_LIST TYPE;
+      };
+
       // demux support
       template <typename TARGET, int N>
       struct DemuxHelper {
diff --git a/runtime/realm/indexspace.h b/runtime/realm/indexspace.h
index 49195963b..c13766a44 100644
--- a/runtime/realm/indexspace.h
+++ b/runtime/realm/indexspace.h
@@ -586,9 +586,10 @@ namespace Realm {
     // would like to use sizeof(IndexSpace<REALM_MAX_DIM, size_t>) here,
     //  but that requires the specializations that are defined in the
     //  include of indexspace.inl below...
-    static const size_t STORAGE_BYTES = (2*REALM_MAX_DIM + 2) * sizeof(size_t);
+    static constexpr size_t MAX_TYPE_SIZE = DIMTYPES::MaxSize::value;
+    static constexpr size_t STORAGE_BYTES = (2*REALM_MAX_DIM + 2) * MAX_TYPE_SIZE;
     typedef char Storage_unaligned[STORAGE_BYTES];
-    REALM_ALIGNED_TYPE_SAMEAS(Storage_aligned, Storage_unaligned, size_t);
+    REALM_ALIGNED_TYPE_SAMEAS(Storage_aligned, Storage_unaligned, DIMTYPES::MaxSizeType<MAX_TYPE_SIZE>::TYPE);
     Storage_aligned raw_storage;
 
   };
```

**Comment by elliottslaughter** [comment]
Builds with Clang 12 now, we'll see what CI says.

**Comment by elliottslaughter** [comment]
Fixed in 1fc45b65f508918ff98430f4831e0e9539de8e14.
